### PR TITLE
[14.x] Pause collection

### PR DESCRIPTION
**Pull request by yaroslawww**
Support of Pausing and Resuming a Subscription.

Subscription api:
```
$subscription = $user->subscription( 'main' );

$resumesAt = Carbon::now()->addWeek();
$subscription->pause(WithPauseCollection::BEHAVIOR_MARK_UNCOLLECTIBLE, $resumesAt);
$subscription->unpause();

$subscription->pauseBehaviorMarkUncollectible($resumesAt);
$subscription->pauseBehaviorKeepAsDraft($resumesAt);
$subscription->pauseBehaviorVoid($resumesAt);

$subscription->syncStripePauseCollection();

$subscription->paused());
$subscription->paused(WithPauseCollection::BEHAVIOR_MARK_UNCOLLECTIBLE);
$subscription->paused(WithPauseCollection::BEHAVIOR_KEEP_AS_DRAFT);
$subscription->paused(WithPauseCollection::BEHAVIOR_VOID);
$subscription->pauseResumesAt();
```

Subscription query api:

```
Subscription::query()->paused()->count();
Subscription::query()->notPaused()->count();

Subscription::query()->paused(WithPauseCollection::BEHAVIOR_VOID)->count();
Subscription::query()->notPaused(WithPauseCollection::BEHAVIOR_VOID)->count();

Subscription::query()->paused(WithPauseCollection::BEHAVIOR_MARK_UNCOLLECTIBLE)->count();
Subscription::query()->notPaused(WithPauseCollection::BEHAVIOR_MARK_UNCOLLECTIBLE)->count();

Subscription::query()->paused(WithPauseCollection::BEHAVIOR_KEEP_AS_DRAFT)->count();
Subscription::query()->notPaused(WithPauseCollection::BEHAVIOR_KEEP_AS_DRAFT)->count();
```

### CODE REVIEW

**Comment by driesvints** [comment]
Thanks, I'll review this when I find some time.

**Comment by driesvints** [comment]
@yaroslawww hey, sorry it took me a while before I got to this. Thanks for all your work on this.

Before I start thoroughly reviewing this can you:

- Remove the `WithPauseCollection` interface. I don't see how that's needed
- Add the constants from the `WithPauseCollection` interface to the Subscription class
- Rename `UsesPauseCollection` to `PausesCollection`
- Revert the factory changes to the user fixture
- Fix all styling issues with StyleCI and use full docblocks everywhere with FQN's
- Remove .idea from the gitignore again (use a global gitignore for this)

Some questions to be answered:

1. Do we really need to keep `behavior` in the database? I think `resumes_at` is enough, no? `behavior` seems to me something you want to know about in your stripe dashboard. Can you give some use cases how you'd use this value in your app?
2. `pause_collection` is currently a JSON array but I'd split that up in two columns: `pause_behavior` & `resumes_at`. Related to the question above I'd only keep `resumes_at` myself. So depending on what we decide for the above we'll need one or two database columns (please wait with making this change until we've discussed this).

**Comment by yaroslawww** [comment]
@driesvints  Sorry for dellay. 
Yes I will refactor code, thank you for feedback.
About question:
1. `behavior` - In my case depending on what behavior used: 
      - app displays different views and messages
      - admin can filter contacts/users including pause behavior using database requests (For example: all contacts with specific interests, specific locations, and pause behavior `keep_as_draft` - then admin export all emails and send marketing offers)
 
    So in this case will be useful to cache the data in database and do not create additional code to call api each time.
2. I created json column to keep this field in case if Stripe will add in future new params in "pause entity". But yes json column add complexity, and for payment module good to keep simplicity. So if you OK with my 1 response then I will add 2 columns?

**Comment by yaroslawww** [comment]
@driesvints I've pushed a new version of the code - please check when you have time.

File: database/migrations/2019_05_03_000004_add_pause_collection_to_subscriptions_table.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+     * Reverse the migrations.
+     */
+    public function down()
+    {
+        Schema::table($this->getTableName(), function (Blueprint $table) {
+            $table->dropColumn('pause_behavior');
+            $table->dropColumn('resumes_at');
+        });
+    }
+}
```
Comment (driesvints) [inline database/migrations/2019_05_03_000004_add_pause_collection_to_subscriptions_table.php:46]: Please add the new `resumes_at` column to the existing migration. Existing Cashier users will need to create this column themselves.

**Comment by yaroslawww** [comment]
@driesvints I added changes

**Comment by driesvints** [comment]
@yaroslawww we're discussing the PR internally atm. I'll get back to you once we reach a decision on how to proceed 👍

**Comment by yaroslawww** [comment]
@driesvints Thank you to feedback. Thanks for the feedback. On weekend I will see what I can do.

**Pull request closed without merging**